### PR TITLE
Update halo with HA 

### PR DESCRIPTION
**Pull request by binbin21avery**
I have changed code with branch lifecycle with HA . I have change code in files halo and halo-cluster 

How to start Halo :
first of all you should git clone kubeblocks-addons and cd to file `addons`

1) helm install etcd with Official etcd of kubeblocks
```helm install etcd ./etcd```
and then helm install etcd-cluster for halo-cluster 
```helm install etcd-cluster ./etcd-cluster```
2) helm install halo with file halo 
if your etcd service is named like first strep, p.s etcd-cluster, now you can helm install halo , if not so , you can edit the file clusterdefinition.yaml ,in file ,which have env named "PATRONI_ETCD3_HOST" .It should change with your own etcd service .
```helm install halo ./halo```
3) helm install halo-cluster with file halo-cluster
```helm install halo-cluster ./halo-cluster```

After all , you can check cluster with order 
```kbcli cluster list ```  

The image of halo is a little big ,so maybe it will take some time to setup cluster , after 60s, the cluster will identity the role in cluster with 3 pods . You can checkout it out with order 

```kbcli cluster halo-cluster list-instances```

But in your k8s must have SC(default)  ,that can dynamic create PV .

### CODE REVIEW

**Comment by binbin21avery** [comment]
here is my kbcli version 
image

**Comment by shanshanying** [comment]
hi @binbin21avery 

please check the script to make sure roleProbe could fetch the correct roles first. Both pods are labled as `replica`, which is unexpected.

**Comment by shanshanying** [comment]
this issue has been solved, and waiting for a update on this PR.

**Comment by shanshanying** [comment]
To run it on KubeBlocks 082, should update the `roleProbl` script as follows:
```bash
- |
  Status=$(curl -s http://localhost:8008) &&
  role=$(echo $Status | jq .role  | tr -d '"') &&
  if [ "$role" = "master" ]; then echo -n "primary"; else echo -n "secondary"; fi
```
return your role with "-n" flag to trim '\n' of returned variable.

File: addons/halo/templates/clusterversion.yaml [hunk 1]
```diff
@@ -2,7 +2,7 @@ apiVersion: apps.kubeblocks.io/v1alpha1
 kind: ClusterVersion
 metadata:
   # major version of the component defined in values.yaml
-  name: halo-{{ .Values.image.tag }}
+  name: halo-14.10
```
[the remark this answers is not in the document]
Comment (binbin21avery) [reply addons/halo/templates/clusterversion.yaml:5]: got it , now I have changed the version in halo-cluster from halo-10.14.231127 to halo-14.10

File: addons/halo/config/restart-parameter.yaml
Comment (shanshanying) [inline addons/halo/config/restart-parameter.yaml:1]: this file is never used?

File: addons/halo-cluster/templates/cluster.yaml [hunk 2]
```diff
@@ -10,4 +10,9 @@
         type: Noop
       {{- include "kblib.componentResources" . | indent 6 }}
       {{- include "kblib.componentStorages" . | indent 6 }}
-      {{- include "kblib.componentServices" . | indent 6 }}
\ No newline at end of file
+      {{- include "kblib.componentServices" . | indent 6 }}
+      serviceRefs:
+      - name: etcdService
+        namespace: default
+        cluster: kb-etcd-for-halo
```
Comment (shanshanying) [inline addons/halo-cluster/templates/cluster.yaml:17]: etcd cluster name is hard-coded? I'd susget use some var defined in `values.yaml' file.

File: addons/halo/templates/configconstraint.yaml [hunk 3]
```diff
@@ -1,19 +1,54 @@
+{{- $cc := .Files.Get "config/halo-config-effect-scope.yaml" | fromYaml }}
 apiVersion: apps.kubeblocks.io/v1alpha1
 kind: ConfigConstraint
 metadata:
   name: halo-cc
   labels:
     {{- include "halo.labels" . | nindent 4 }}
 spec:
+  reloadOptions:
```
Comment (shanshanying) [inline addons/halo/templates/configconstraint.yaml:9]: there is no `dynamicParameter` defined , there is no need to config `reloadOptions`, which is to trigger dynamic parameters wthout RESTART. Just remove it.

File: addons/halo/templates/clusterdefinition.yaml [hunk 4]
```diff
@@ -14,8 +14,42 @@ spec:
...
+            - image: {{ .Values.image.registry }}/{{ .Values.image.repository}}:{{.Values.image.tag}}
+              command: 
+                - |
+                  Status=$(curl -s http://localhost:8008) && 
+                  role=$(echo $Status | jq .role  | tr -d '"') &&
+                  if [ "$role" = "master" ]; then echo  "primary"; else echo  "secondary"; fi
+          roleUpdateMechanism: DirectAPIServerEventUpdate
+          initialDelaySeconds: 60
+        memberUpdateStrategy: BestEffortParallel
+      serviceRefDeclarations:
```
Comment (shanshanying) [inline addons/halo/templates/clusterdefinition.yaml:48]: serviceRef is never used. ETCD address is injected by ENV `etcd-service`.  FYI. You may refer to doc for more info on serviceRefs. https://cn.kubeblocks.io/docs/preview/developer-docs/external-component/reference-external-component

File: addons/halo/config/restart-parameter.yaml
Comment (binbin21avery) [reply addons/halo/config/restart-parameter.yaml:1]: now already delete

File: addons/halo-cluster/templates/cluster.yaml (hunk 2 above)
Comment (binbin21avery) [reply addons/halo-cluster/templates/cluster.yaml]: now I have changed the Cluster architecture of halo-cluster , independence the service etcd

File: addons/halo/templates/configconstraint.yaml (hunk 3 above)
Comment (binbin21avery) [reply addons/halo/templates/configconstraint.yaml]: got it

**Comment by shanshanying** [comment]
LGTM

**Review by wangyelei — APPROVED** [review]

**Review by shanshanying — APPROVED** [review]

**Pull request merged**